src/main/java/edu/kit/compiler/optimizations/CallGraph.java

```diff
@@ -0,0 +1,234 @@
+package edu.kit.compiler.optimizations;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.jgrapht.Graph;
+import org.jgrapht.Graphs;
+import org.jgrapht.alg.connectivity.KosarajuStrongConnectivityInspector;
+import org.jgrapht.graph.DefaultDirectedWeightedGraph;
+import org.jgrapht.graph.DefaultEdge;
+
+import firm.Entity;
+import firm.Program;
+import firm.bindings.binding_irnode.ir_opcode;
+import firm.nodes.Address;
+import firm.nodes.Call;
+import firm.nodes.NodeVisitor;
+import lombok.AccessLevel;
+import lombok.AllArgsConstructor;
+import lombok.RequiredArgsConstructor;
+
+/**
+ * Represents the call graph of a Firm program.
+ * 
+ * Note: Whenever Entities are passed as parameters or returned from any method
+ * in this class, it is assumed that they represent a function. Entities are NOT
+ * required to have an associated graph, e.g. external function, like those of
+ * the standard library are part of the call graph.
+ * Beware that there are no guarantees that entities without associated graph
+ * will be represented in the call graph (see also `#create()`).
+ */
+@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
+public final class CallGraph {
+
+    private final Graph<Entity, DefaultEdge> graph;
+
+    private Components components;
+
+    /**
+     * Return a stream containing every function that may call the given function.
+     */
+    public Stream<Entity> getCallers(Entity function) {
+        return graph.incomingEdgesOf(function).stream()
+                .map(graph::getEdgeSource);
+    }
+
+    /**
+     * Return a stream containing every function that may be called by the given
+     * function.
+     */
+    public Stream<Entity> getCallees(Entity function) {
+        return graph.outgoingEdgesOf(function).stream()
+                .map(graph::getEdgeTarget);
+    }
+
+    /**
+     * Return the number of distinct calls from caller to callee.
+     */
+    public double getCallFrequency(Entity caller, Entity callee) {
+        var edge = graph.getEdge(caller, callee);
+        return edge == null ? 0.0 : graph.getEdgeWeight(edge);
+    }
+
+    /**
+     * Return true if `caller` may directly call `callee`.
+     */
+    public boolean existsCall(Entity caller, Entity callee) {
+        return graph.containsEdge(caller, callee);
+    }
+
+    /**
+     * Returns true if there may exist a recursion if caller calls callee,
+     * i.e. there exists some path through the call graph such that caller
+     * (in-)directly calls itself.
+     */
+    public boolean existsRecursion(Entity caller, Entity callee) {
+        // ? what should this fn do if caller does not call callee
+        // ? currently returns false, maybe throw exception instead
```

Comment:
I think returning `false` is alright. If the user of this function is interested in knowing whether a call exists at all, they should call `existsCall` in addition.